drivers/mspi/mspi_ambiq_ap4.c

```diff
@@ -0,0 +1,1530 @@
+			ret = -EHOSTDOWN;
+			goto e_return;
+		}
+
+		// if (hal_dev_cfg.eDeviceConfig != AM_HAL_MSPI_FLASH_SERIAL_CE0) {
+		// 	ret = pinctrl_apply_state(cfg->pcfg,
+		// 				  PINCTRL_STATE_PRIV_START + dev_id->dev_idx);
+		// 	if (ret) {
+		// 		goto e_return;
+		// 	}
+		// }
+
+		data->dev_cfg = *dev_cfg;
+		data->dev_id = (struct mspi_dev_id *)dev_id;
+	}
+
+	data->hal_dev_cfg = hal_dev_cfg;
+
+	return ret;
+
+e_return:
+	k_mutex_unlock(&data->lock);
+	return ret;
+}
+
+static int mspi_ambiq_xip_config(const struct device *controller, const struct mspi_dev_id *dev_id,
+				 const struct mspi_xip_cfg *xip_cfg)
+{
+	const struct mspi_ambiq_config *cfg = controller->config;
+	struct mspi_ambiq_data *data = controller->data;
+	am_hal_mspi_request_e eRequest;
+	int ret = 0;
+
+	if (dev_id != data->dev_id) {
+		LOG_INST_ERR(cfg->log, "%u, dev_id don't match.", __LINE__);
+		return -ESTALE;
+	}
+
+	ret = mspi_check_mem_size(xip_cfg->size);
+
+	if (ret > ((int)AM_HAL_MSPI_AP_SIZE_MAX)) {
+		LOG_INST_ERR(cfg->log, "%u, fail to configure eAPSize.", __LINE__);
+		return -EHOSTDOWN;
+	}
+
+	data->hal_xip_cfg.eAPSize = mspi_get_mem_apsize(ret);
+
+	if (xip_cfg->enable) {
+		eRequest = AM_HAL_MSPI_REQ_XIP_EN;
+	} else {
+		eRequest = AM_HAL_MSPI_REQ_XIP_DIS;
+	}
+
+	ret = am_hal_mspi_control(data->mspiHandle, AM_HAL_MSPI_REQ_XIP_CONFIG, &data->hal_xip_cfg);
+	if (ret) {
+		LOG_INST_ERR(cfg->log, "%u, fail to configure XIP REQ config, code:%d.", __LINE__,
+			     ret);
+		return -EHOSTDOWN;
+	}
+
+	ret = am_hal_mspi_control(data->mspiHandle, AM_HAL_MSPI_REQ_XIP_MISC_CONFIG,
+				  &data->hal_xip_misc_cfg);
+	if (ret) {
+		LOG_INST_ERR(cfg->log, "%u, fail to configure XIP MISC config, code:%d.", __LINE__,
+			     ret);
+		return -EHOSTDOWN;
+	}
+
+	ret = am_hal_mspi_control(data->mspiHandle, eRequest, NULL);
+	if (ret) {
+		LOG_INST_ERR(cfg->log, "%u,Unable to complete xip config:%d.", __LINE__,
+			     xip_cfg->enable);
+		return -EHOSTDOWN;
+	}
+
+	data->xip_cfg = *xip_cfg;
+	return ret;
+}
+
+static int mspi_ambiq_scramble_config(const struct device *controller,
+				      const struct mspi_dev_id *dev_id,
+				      const struct mspi_scramble_cfg *scramble_cfg)
+{
+	const struct mspi_ambiq_config *cfg = controller->config;
+	struct mspi_ambiq_data *data = controller->data;
+	am_hal_mspi_xip_config_t hal_xip_cfg = data->hal_xip_cfg;
+	am_hal_mspi_request_e eRequest;
+	int ret = 0;
+
+	if (mspi_is_inp(controller)) {
+		return -EBUSY;
+	}
+
+	if (dev_id != data->dev_id) {
+		LOG_INST_ERR(cfg->log, "%u, dev_id don't match.", __LINE__);
+		return -ESTALE;
+	}
+
+	if (scramble_cfg->enable) {
+		eRequest = AM_HAL_MSPI_REQ_SCRAMB_EN;
+	} else {
+		eRequest = AM_HAL_MSPI_REQ_SCRAMB_DIS;
+	}
+
+	ret = am_hal_mspi_disable(data->mspiHandle);
+	if (ret) {
+		LOG_INST_ERR(cfg->log, "%u, fail to disable MSPI, code:%d.", __LINE__, ret);
+		return -EHOSTDOWN;
+	}
+
+	ret = am_hal_mspi_control(data->mspiHandle, eRequest, NULL);
```

Comment:
You should put this behind AM_HAL_MSPI_REQ_SCRAMBLE_CONFIG/AM_HAL_MSPI_REQ_XIP_CONFIG